src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +113,252 @@ object RuleUtils {
+          .toSet -- index.allSourceFileInfos).toSeq.map(f => new Path(f.name))
+
+        if (filesAppended.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
+    }
+
+    if (!originalPlan.equals(complementIndexPlan)) {
+      // Remove sort order because we cannot guarantee the ordering of source files
+      val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+
+      object ExtractTopLevelPlanForShuffle {
+        type returnType = (LogicalPlan, Seq[Option[Attribute]], Boolean)
+        def unapply(plan: LogicalPlan): Option[returnType] = plan match {
+          case project @ Project(
+                _,
+                Filter(_, LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _))) =>
+            val indexedAttrs = getIndexedAttrs(project, index.indexedColumns)
+            Some(project, indexedAttrs, true)
+          case project @ Project(
+                _,
+                LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _)) =>
+            val indexedAttrs = getIndexedAttrs(project, index.indexedColumns)
+            Some(project, indexedAttrs, true)
+          case filter @ Filter(_, LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _)) =>
+            val indexedAttrs = getIndexedAttrs(filter, index.indexedColumns)
+            Some(filter, indexedAttrs, false)
+          case relation @ LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _) =>
+            val indexedAttrs = getIndexedAttrs(relation, index.indexedColumns)
+            Some(relation, indexedAttrs, false)
+        }
+        def getIndexedAttrs(
+            plan: LogicalPlan,
+            indexedColumns: Seq[String]): Seq[Option[Attribute]] = {
+          val attrMap = plan.output.attrs.map(attr => (attr.name, attr)).toMap
+          indexedColumns.map(colName => attrMap.get(colName))
+        }
+      }
+
+      // 2) Perform on-the-fly Shuffle with the same partition structure of index
+      // so that we could avoid incurring Shuffle of whole index data at merge stage.
+      // In order to utilize push-down filters, we would locate Shuffle node after
+      // Project or Filter node. (Case 1)
+      // However, if Project node excludes any of indexedColumns, Shuffle will be
+      // converted to RoundRobinPartitioning which can cause wrong result issues.
+      // So Shuffle should be located before Project node in that case. (Case 2)
+      // Case 1) Shuffle => Project => Filter => Relation (Project&Filter will be pushed down)
+      // Case 2) Project => Shuffle => Filter => Relation (Filter will be pushed down)
```

Comment:
I am pretty sure I made this change but don't know what happened to that suggestion :) Sorry I made it a bit verbose (please feel free to modify as you see fit). I just thought this is such an important piece that it needs some explanation so the "future me" won't forget :D 

Also, note that I reversed Case 1 and 2 to write this properly. 

```suggestion
      // Step 2) During merge of the index data with the newly appended data, 
      // perform on-the-fly shuffle of the appended-data with the same partition 
      // structure of index so that we could avoid incurring shuffle of the index 
      // data (which is assumed to be the larger sized dataset).
      // 
      // To do this optimally, we would have to push-down filters/projects before
      // shuffling the newly appended data to reduce the amount of data that gets 
      // shuffled. However, if project excludes any of the [[indexedColumns]], Spark 
      // handles the shuffle through [[RoundRobinPartitioning]], which can cause 
      // wrong results.
      //
      // Therefore, we do the following:
      // if Project node excludes any of the [[indexedColumns]]:
      //    Case 1: Shuffle should be located **before** Project but **after** Filter
      //                Plan: Project => Shuffle => Filter => Relation 
      // else:
      //    Case 2: Shuffle node should come **after** Project and/or Filter node 
      //                Plan: Shuffle => Project => Filter => Relation
```

Reply:
I updated the description but we might not need to handle Case1) because `Shuffle` is only applied for Join Index which requires all indexed columns in its join condition. (refer https://github.com/microsoft/hyperspace/pull/165#discussion_r490674500)